### ISSUE TITLE
Exploring wasm options

### ISSUE THREAD

**Issue by VarLad** [JuliaWGPU/WGPUgfx.jl#15]
Posting some links here to look into them later. (and for others if interested)
Wasm + Julia have come a far way. 
It would be ideal to have the JuliaWGPU stack on wasm (as the goal is to be able to run everything on the web to begin with)

- [WebAssemblyCompiler.jl](https://github.com/tshort/WebAssemblyCompiler.jl) : Its docs show many cool things, including Makie plot running in the browser
- [Wasmtime.jl](https://github.com/Pangoraw/Wasmtime.jl) : To effortlessly execute wasm code from Julia.
- [An example notebook by Pangoraw](https://www.ber.gp/notebooks/fun%20theory)

**Comment by arhik** [comment]
I completely agree. `WebAssemblyCompiler` looks promising. While `WebAssemblyCompiler` is using SSA based approach, I will be open to any library that can do this compilation. I tried LLVM based approach here https://github.com/arhik/WASMCompiler.jl. it works. I was able to compile julia functions with LLVM tools (look into target.jl for hints) and do basic calculations on browser. And also, some arrays processing. I couldn't trigger `console log` and print strings though. I also had difficulties with 64-bit pointers and 32-bit pointer conversions. If `WebAssemblyCompiler.jl` can do those seamlessly we can ditch LLVM based approach.  WASM now has garbage collection feature added. Not sure what it means for us but all this hints us it is good time to work on basic examples for `JuliaWGPU` on browser.

**Comment by tshort** [comment]
Having JuliaWGPU would be awesome. If Makie could be tied in, that'd be even better. Let me know if you run into issues with WebAssemblyCompiler.

**Comment by VarLad** [comment]
@tshort Thanks for the answer. 
What would be needed to support:
- Pointers
- Union Types
- Integer types
- C deps

in WebAssemblyCompiler.jl ?
Also, Memory type PR has been merged.

**Comment by tshort** [comment]
* *Pointers*--This isn't possible currently. Maybe some sort of PseudoPtr would work, depending on specific needs. Future additions to WASM-GC may cover more use cases.
* *Union Types*--The easiest way is probably to create a type that can hold all possibilities along with a type flag. That's not very memory efficient, but I'm not sure WASM-GC can handle more complicated reinterpretation.
* *Integer types*--Int128 and similar is probably the main need. That's best to emulate on the Julia side and swap in where needed.
* *C deps*--This should work now. Compile a library to WebAssembly with Emscripten. `ccall` should just work. Either link the wasm files, or pull them in separately. I don't think that Emscripten/LLVM can handle GC'd arrays and objects, though, so common things like OpenBLAS won't work (unless you copy the GC'd array to wasm's linear memory).
